### ISSUE TITLE
Proposal: Explicit inclusion rewriting

### ISSUE THREAD

**Issue by psalz** [illuhad/hipSYCL#20]
I propose adding a special flag for `syclcc` which controls whether an include path should be rewritten or not. Here’s why:

First of all I suspect that for many smaller SYCL applications, all kernel code will be situated within implementation files anyway, so the entire rewrite pass could be omitted, improving compilation speeds.
More importantly though, inlining of headers can sometimes cause compilation errors. As an example, the following program does not compile with syclcc (using nvcc, as of Boost 1.68):

```c++
#include <boost/optional.hpp>

int main() {
    boost::optional<int> foo, bar;
    foo = bar;
    return 0;
}
```

If I prevent the inclusion rewriter from inlining Boost (by adding a condition within `RewriteSelector::shouldIncludeBeRewritten`), everything works fine. I can’t tell you the exact reason, as Boost is huge and contains a metric ton of workarounds for various compilers (I really don’t envy those guys!), however, to me this looks like a case of compiler detection which causes the compilation to take different branches during the inclusion rewriting and the compilation with `nvcc`. I actually have no idea how they do it, I tried to create a small test case to replicate this and couldn’t come up with one. Nevertheless, it apparently does happen.

Moreover, all of the weird egde-cases I’ve encountered with pruning were within libraries that don’t contain any device code at all. Omitting them from the rewrite pass would also alleviate those problems.

Lastly, this could generally improve compilation performance, as we would have less text to dump back to disk.

---

So, instead of rewriting everything but hipSYCL and system libraries, I propose that headers which should be rewritten be made explicit through some flag for syclcc, e.g. `-IR` or `--rewrite-includes-path=`. I suspect that in many cases the rewrite step would only concern local includes (e.g. `-IR./include`) or could even be omitted entirely. Also maintaining a list of system library headers to exclude wouldn’t be required anymore either.

**Comment by illuhad** [comment]
I agree with the general strategy that we should try to only fiddle with code that we need to, and ignore irrelevant parts.
That said, I would like to at least in principle support SYCL header-only libraries (like parallel STL) that are installed somewhere in the system without additional effort from the user. So, I would prefer if we still rewrite everything by default. But we could have additional flags, as you propose, to control this. I.e.
* By default everything is rewritten
* if specific rewrite include paths are passed, only these are rewritten.
The name of this argument should probably be something that the user can immediately understand, like `--device-code-path` or something like that, since this option would effectively control where kernel code is allowed.

Out of curiosity: Does the boost issue also pop up if we change the compiler identification during the rewrite/transformation step to whatever nvcc is using? Especially with boost we will likely get other weird problems because of that but I wonder if this could be a solution for cases when we cannot use the installed clang as nvcc's default host compiler (it's very picky about the host compilers it supports, even cuda 10 only supports clang 6, nothing newer or older).

**Comment by psalz** [comment]
That's actually a great compromise! Also in hindsight I agree that `--rewrite-includes-path` is a terrible name from a user's perspective, heh. I'd however suggest to still make clear that it's about includes, e.g. `--device-includes-path`.

Regarding your question: I went back and tried setting e.g. `-D__NVCC__` and `-U__clang__` on the rewrite pass but that just causes more problems - I guess Boost is simply too complex. I also realized that I already had set Clang as the nvcc host compiler and tried it again with gcc. It also fails, but apparently due to a different error ;-)

**Comment by illuhad** [comment]
I wouldn't be surprised if there were generally some parts of boost that nvcc simply is incapable of compiling (although not in this particular case since you say that it works if boost is not inlined). Usually nvcc is only used to compile *.cu files which typically don't include boost, and nvcc is a huge hack in itself. Anyway, thanks for trying :)

As for the name: Just collecting ideas:
* `--device-includes-path`
* `--device-header-path`
* `--allowed-device-header-path` (bit long, but most clear?)

I think I prefer "header" over "includes" since it sounds less ambiguous to me.

If we effectively don't allow kernel code in certain paths anymore, we should also think about whether there is a way to diagnose that, i.e. present the user with a meaningful error if he has kernel code in a "forbidden" location. Otherwise, the user would only see nvcc/clang errors that `__device__` or `__host__` attributes are missing due to the missing rewrite step which doesn't really point to the issue... This could be done by checking in `hipsycl_transform_source` at the moment of writing these attributes if the buffer that we're writing to belongs to the main input file. If that's not the case, we are in a file that was not rewritten which should then cause an error (or warning?) to be emitted.
But we can implement this as a second step, and first implement selective rewrites before worrying about diagnostics :)

**Comment by illuhad** [comment]
Implemented in PR #33.
